code/modules/reagents/Chemistry-Holder.dm

```diff
@@ -103,162 +73,138 @@
-	return 0
+			newreagent.set_temperature(temperature, src, TRUE)
+		if(!thermal_energy && temperature != newreagent.get_temperature(src))
+			crash_with("Temperature [temperature] did not match [newreagent.get_temperature(src)] for NEW reagent [newreagent.type]!")
+	else	// Existing reagent
+		var/old_energy = newreagent.get_thermal_energy(src)
+		reagent_volumes[rtype] += amount
+		if(!isnull(data))
+			LAZYSET(reagent_data, rtype, newreagent.mix_data(src, data, amount))
+		if(temperature <= 0)
+			temperature = newreagent.default_temperature
+		newreagent.add_thermal_energy(old_energy, src, TRUE) // This part has the safety var set because thermal shock shouldn't occur due to it.
+		if(thermal_energy > 0) // This if-else is for the change from the current temperature.
+			newreagent.add_thermal_energy(thermal_energy - old_energy, src, FALSE)
+		else
+			newreagent.add_thermal_energy(newreagent.default_temperature*newreagent.specific_heat*amount - old_energy, src, FALSE)
+		if(!thermal_energy && round(temperature, 1) != round(newreagent.get_temperature(src), 1))
+			crash_with("Temperature [temperature] did not match [newreagent.get_temperature(src)] for EXISTING reagent [newreagent.type]!")
+	UNSETEMPTY(reagent_volumes)
+	update_holder(!safety)
+	return TRUE
 
 /datum/reagents/proc/remove_reagent(var/rtype, var/amount, var/safety = 0)
-	if(!isnum(amount))
-		return 0
-
-	for(var/datum/reagent/current in reagent_list)
-		if(current.type == rtype)
-			amount = min(amount,current.volume)
-			var/old_volume = current.volume
-			current.volume -= amount
-			current.add_thermal_energy( -(current.thermal_energy * (amount/old_volume)) )
-			update_holder(!safety)
-			return 1
-	return 0
+	var/old_volume = REAGENT_VOLUME(src, rtype)
+	if(!isnum(amount) || old_volume <= 0)
+		return FALSE
+	amount = min(amount, old_volume)
+	reagent_volumes[rtype] -= amount
+	var/decl/reagent/current = decls_repository.get_decl(rtype)
+	current.add_thermal_energy( -(current.get_thermal_energy(src) * (amount/old_volume)), src )
+	update_holder(!safety)
+	return TRUE
 
 /datum/reagents/proc/del_reagent(var/rtype)
-	for(var/datum/reagent/current in reagent_list)
-		if (current.type == rtype)
-			if(ismob(my_atom))
-				current.final_effect(my_atom)
-			reagent_list -= current
-			qdel(current)
-			update_holder(FALSE)
-			return 0
+	if(REAGENT_VOLUME(src, rtype) <= 0)
+		return FALSE
+	var/decl/reagent/current = decls_repository.get_decl(rtype)
+	if(ismob(my_atom))
+		current.final_effect(my_atom, src)
+	LAZYREMOVE(reagent_data, rtype)
+	LAZYREMOVE(reagent_volumes, rtype)
+	if(primary_reagent == rtype)
+		primary_reagent = null
+	update_holder(FALSE)
+	return FALSE
 
 /datum/reagents/proc/has_reagent(var/rtype, var/amount = 0)
-	for(var/datum/reagent/current in reagent_list)
-		if(current.type == rtype)
-			if(current.volume >= amount)
-				return 1
-			else
-				return 0
-	return 0
+	return (rtype in reagent_volumes) && reagent_volumes[rtype] >= amount
```

Comment:
```suggestion
	return REAGENT_VOLUME(src, rtype) >= amount
```

Reply:
NO! If `REAGENT_VOLUME(src, rtype) == null`, i.e. if rtype is not in reagent_volumes, and you use the default argument of 0, you get `null >= 0` which is TRUE. It should instead be FALSE, which requires this check. https://i.imgur.com/UzNP55J.png

Reply:
(This change was made because I initially used `REAGENT_VOLUME` and it failed horribly.)

Reply:
Makes sense.

Reply:
Makes sense.